### PR TITLE
Issue 1865: Added profiles to enforce JDK 1.8.0-241 or 11.3 and up.

### LINKED ISSUE #1865: Add a rule using the maven-enforcer plugin that requires JDK 8 (or 11)
# Task Description

We should add a Maven Enforcer rule that requires:
* JDK 1.8 (if possible, `1.8.0_241` or higher) in order to build the code. This check should be in a Maven profile that's triggered, unless the JDK is 11.x.
* JDK 11 (if possible, `11.3`, or a higher `11.x`) in order to build the code. This check should be in a Maven profile that's triggerred only if the JDK is not JDK 1.8.

Please, note that the code in the `master` does not currently compile using JDK11. If the enforcer rule can be added for this, then this will be pre-work for JDK11 support. (You may want to have a look at the tasks related to #1111).

# Tasks

The following tasks will need to be carried out:
* [ ] Add this rule in the [strongbox-parent].
* [ ] Apply this rule to the [strongbox] project.
* [ ] Apply this rule to the [strongbox-web-integration-tests].

# Task Relationships

This task:
* Is relates to #1111 , #1864

# Help

* [Our chat](https://chat.carlspring.org/)
* Points of contact:
  * @carlspring
  * @sbespalov
  * @steve-todorov

[strongbox]: https://github.com/strongbox/strongbox/
[strongbox-parent]: https://github.com/strongbox/strongbox-parent/
[strongbox-web-integration-tests]: https://github.com/strongbox/strongbox-web-integration-tests/

### PR DESCRIPTION
**Pull request by AricMonary**
# Pull Request Description

This pull request closes #1865

# Acceptance Test

* [X] Building the code with `mvn clean install -Dintegration.tests` still works.
* [X] Running `mvn spring-boot:run` in the `strongbox-web-core` still starts up the application correctly.
* [X] Building the code and running the `strongbox-distribution` from a `zip` or `tar.gz` still works.
* [X] The tests in the [`strongbox-web-integration-tests`](https://github.com/strongbox/strongbox-web-integration-tests/) still run properly.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [X] No

* Does this pull request require other pull requests to be merged first? 
  * [ ] Yes, please see #...
  * [X] No

* Does this require an update of the documentation?
  * [ ] Yes, please see strongbox/strongbox-docs#{PR_NUMBER}
  * [X] No

### CODE REVIEW

**Comment by AricMonary** [comment]
Working on a fix. Seems that the tests use a lower level of java than wanted. Gonna add maven enforcement too.

File: pom.xml [hunk 1]
```diff
@@ -119,13 +119,77 @@
                                 <version>8.29</version>
                             </dependency>
                         </dependencies>
-                        <configuration>
+                        <configuration>
```
Comment (carlspring) [inline pom.xml:122]: Please, restore the original formatting.
```suggestion
                        <configuration>
```

File: pom.xml [hunk 2]
```diff
@@ -119,13 +119,77 @@
...
+                        <configuration>     
                             <configLocation>strongbox-checks.xml</configLocation>
                             <consoleOutput>true</consoleOutput>
                         </configuration>
                     </plugin>
                 </plugins>
             </build>
         </profile>
+        <profile>
+            <id>Enforce-Java-11</id>
```
Comment (carlspring) [inline pom.xml:131]: Please, change this to:
```suggestion
            <id>enforce-jdk11</id>
```

File: pom.xml [hunk 3]
```diff
@@ -119,13 +119,77 @@
...
+                                    </rules>
+                                </configuration>
+                            </execution>
+                        </executions>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
+        <profile>
+            <id>Enforce-Java-8</id>
```
Comment (carlspring) [inline pom.xml:163]: Please, change this to:
```suggestion
            <id>enforce-jdk8</id>
```

File: pom.xml [hunk 4]
```diff
@@ -119,13 +119,77 @@
...
+            <activation>
+                <jdk>1.8</jdk>
+            </activation>
+            <build>
+                <plugins>
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-enforcer-plugin</artifactId>
+                        <version>3.0.0-M3</version>
```
Comment (carlspring) [inline pom.xml:172]: As explained in the other pull request, please move the `version` definition into the `strongbox-parent`'s `<pluginManagement/>` section. Thanks! :)

File: pom.xml [hunk 5]
```diff
@@ -119,13 +119,77 @@
...
+                            <execution>
+                                <id>enforce-versions</id>
+                                <goals>
+                                    <goal>enforce</goal>
+                                </goals>
+                                <configuration>
+                                    <rules>
+                                        <requireJavaVersion>
+                                            <version>[1.8,)</version>
+                                            <message>Strongbox only supports JDK 1.8.0-241 and up.</message>
```
Comment (carlspring) [inline pom.xml:183]: Please, change this to:
```suggestion
                                            <message>Strongbox only supports JDK 1.8.0-241, or higher.</message>
```

File: pom.xml [hunk 6]
```diff
@@ -119,13 +119,77 @@
...
+            <id>Enforce-Java-11</id>
+            <activation>
+                <jdk>11</jdk>
+            </activation>
+            <build>
+                <plugins>
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-enforcer-plugin</artifactId>
+                        <!--<version>3.0.0-M3</version>-->
```
Comment (carlspring) [inline pom.xml:140]: Please, remove this.

File: pom.xml (hunk 3 above)
Comment (AricMonary) [reply pom.xml:163]: resolved

File: pom.xml (hunk 2 above)
Comment (AricMonary) [reply pom.xml:131]: resolved

File: pom.xml (hunk 1 above)
Comment (AricMonary) [reply pom.xml:122]: resolved

File: pom.xml [hunk 7]
```diff
@@ -119,13 +119,77 @@
...
+                            <execution>
+                                <id>enforce-versions</id>
+                                <goals>
+                                    <goal>enforce</goal>
+                                </goals>
+                                <configuration>
+                                    <rules>
+                                        <requireJavaVersion>
+                                            <version>[11.3,)</version>
+                                            <message>Strongbox only supports JDK 11.3 and up.</message>
```
[the remark this answers is not in the document]
Comment (AricMonary) [reply pom.xml:151]: resolved

File: pom.xml (hunk 6 above)
Comment (AricMonary) [reply pom.xml:140]: resolved

File: pom.xml (hunk 5 above)
Comment (AricMonary) [reply pom.xml:183]: resolved

**Comment by carlspring** [comment]
There should actually not be a need for this pull request. When you apply the changes to the parent, these should be propagated to any projects that extend it. Do you have the Maven profiles defined there? Could you test, if this works?

Thanks! :)

File: pom.xml (hunk 4 above)
Comment (AricMonary) [reply pom.xml:172]: resolved

**Pull request closed without merging**